crates/sui-core/src/authority/submitted_transaction_cache.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use consensus_types::block::Round;
+use parking_lot::RwLock;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::net::IpAddr;
+use sui_types::digests::TransactionDigest;
+use sui_types::traffic_control::Weight;
+use tracing::debug;
+
+/// The number of consensus rounds to retain transaction submission information before garbage collection.
+pub(crate) const SUBMISSION_RETENTION_ROUNDS: u32 = 400;
+
+/// TODO(fastpath): Evaluate if this should be a configurable parameter.
+/// The allowed number of retries allowed regardless of gas price.
+pub(crate) const DEFAULT_RETRY_TOLERANCE: u32 = 20;
+
+/// Cache for tracking submitted transactions to prevent DoS through excessive resubmissions.
+/// Tracks submission counts and enforces gas-price-based amplification + retry tolerance limits.
+pub(crate) struct SubmittedTransactionCache {
+    inner: RwLock<Inner>,
+    retention_rounds: u32,
+    retry_tolerance: u32,
+}
+
+#[derive(Default)]
+struct Inner {
+    /// Map of transaction digest to submission metadata
+    transactions: HashMap<TransactionDigest, SubmissionMetadata>,
+    /// Transactions indexed by submitted round for efficient GC
+    transactions_by_round: BTreeMap<Round, HashSet<TransactionDigest>>,
+    /// The last committed leader round for GC purposes
+    last_committed_round: Option<Round>,
+}
+
+#[derive(Debug, Clone)]
+struct SubmissionMetadata {
+    /// Round when transaction was last seen (for GC)
+    submitted_round: Round,
+    /// Number of times this transaction has been submitted
+    submission_count: u32,
+    /// Maximum allowed submissions based on gas price amplification + retry tolerance
+    max_allowed_submissions: u32,
+    /// Client IP address that submitted this transaction
+    submitter_client_addr: Option<IpAddr>,
+}
+
+impl SubmittedTransactionCache {
+    pub(crate) fn new(retention_rounds: Option<u32>, retry_tolerance: Option<u32>) -> Self {
+        let retention_rounds = retention_rounds.unwrap_or(SUBMISSION_RETENTION_ROUNDS);
+
+        Self {
+            inner: Default::default(),
+            retention_rounds,
+            retry_tolerance: retry_tolerance.unwrap_or(DEFAULT_RETRY_TOLERANCE),
+        }
+    }
+
+    pub(crate) fn record_submitted_tx(
+        &self,
+        digest: &TransactionDigest,
+        submitted_round: Round,
+        amplification_factor: u32,
+        submitter_client_addr: Option<IpAddr>,
+    ) {
+        let mut inner = self.inner.write();
+
+        // We allow 1 submission for the initial submission, and then the retry tolerance + amplification factor
+        // for subsequent submissions.
+        let max_allowed_submissions = 1 + self.retry_tolerance + amplification_factor;
```

Comment:
The extra `1` should be unnecessary. [1, 2) * RGP allows 1 submission, [2, 3) * RGP allows 2 submissions etc.